### PR TITLE
MM-9779: Incorporate a Token into the invitations system

### PR DESCRIPTION
**Pull request by jespino**
#### Summary
It binds each invitation to a token, avoiding the invitations re-use.

#### Ticket Link
[MM-9779](https://mattermost.atlassian.net/browse/MM-9779)
PR for the API documentation mattermost/mattermost-api-reference#349

#### Checklist
- [x] Added or updated unit tests (required for all new features)
- [x] Added API documentation (required for all new APIs)
- [x] All new/modified APIs include changes to the drivers

### CODE REVIEW

File: app/team.go [hunk 1]
```diff
@@ -223,8 +227,17 @@ func (a *App) AddUserToTeamByHash(userId string, hash string, data string) (*mod
...
+	if result.Err != nil {
+		return nil, model.NewAppError("JoinUserToTeamByHash", "api.user.create_user.signup_link_invalid.app_error", nil, result.Err.Error(), http.StatusBadRequest)
+	}
+
+	token := result.Data.(*model.Token)
+	if token.Type != TOKEN_TYPE_TEAM_INVITATION {
+		return nil, model.NewAppError("JoinUserToTeamByHash", "api.user.create_user.signup_link_invalid.app_error", nil, "", http.StatusBadRequest)
+	}
+
+	if model.GetMillis()-token.CreateAt >= TEAM_INVITATION_EXPIRY_TIME {
```
Comment (crspeller) [inline app/team.go]: Expired tokens won't be deleted. Same if there are any errors below. (defer the token deletion?)

File: app/team.go [hunk 2]
```diff
@@ -223,8 +227,17 @@ func (a *App) AddUserToTeamByHash(userId string, hash string, data string) (*mod
 		return nil, model.NewAppError("JoinUserToTeamByHash", "api.user.create_user.signup_link_invalid.app_error", nil, "", http.StatusBadRequest)
```
Comment (crspeller) [inline app/team.go]: Why are we still doing hash verification? We should just use the token if we are going to add that.

File: app/user.go [hunk 3]
```diff
@@ -77,6 +89,10 @@ func (a *App) CreateUserWithHash(user *model.User, hash string, data string) (*m
 
 	a.AddDirectChannels(team.Id, ruser)
 
+	if err := a.DeleteToken(token); err != nil {
```
Comment (crspeller) [inline app/user.go]: Again problems above will cause the token not to be deleted.

File: app/team.go (hunk 2 above)
Comment (jespino) [reply app/team.go]: We still need to send data like the email (shared with the webapp) so I wanted to maintain the changes small. Besides the quantity of extra data that can be included in the token is only 128 bytes, so we can't include all the current data in the token.

File: app/team.go (hunk 1 above)
Comment (jespino) [reply app/team.go]: It makes sense to remove the expired tokens, but I though that was already done by another gorouting periodically. There are other tokens that can be expired (reset password, email verification). You can ask for a reset password, and never reset it. If we have to delete expired tokens I prefer to do that as a separated block of code in another PR.

File: app/user.go (hunk 3 above)
Comment (jespino) [reply app/user.go]: If there is any problem above, you don't want to get deleted the token, you only want to delete (invalidate) the token, after the proper use. Any fail try not invalidate the token.

File: app/team.go (hunk 1 above)
Comment (crspeller) [reply app/team.go]: Ya the tokens are cleaned up automatically, but I just thought you might want to be better about it here.

File: app/team.go (hunk 2 above)
Comment (crspeller) [reply app/team.go]: I would rather remove the extra complexity. You don't need to have verified all the data that we currently send down. Only the stuff we actually trust needs to be stored in the token. (so userId and teamId, the rest is just for show)

File: app/team.go (hunk 1 above)
Comment (jespino) [reply app/team.go]: Well, I can delete here explicitly the deleted token, but, as in the other case we only want to delete the token if it is used (or like in this cases, if it is expired.

File: app/team.go (hunk 2 above)
Comment (jespino) [reply app/team.go]: Yes, you are right, probably we don't need that complexity.

**Comment by crspeller** [comment]
@jespino You have some related unit tests breaking.

**Comment by jespino** [comment]
Now is ready with test updated :)

File: app/email.go [hunk 4]
```diff
@@ -270,15 +270,21 @@ func (a *App) SendInviteEmails(team *model.Team, senderName string, invites []st
...
-			props["id"] = team.Id
 			props["display_name"] = team.DisplayName
 			props["name"] = team.Name
-			props["time"] = fmt.Sprintf("%v", model.GetMillis())
 			data := model.MapToJson(props)
-			hash := utils.HashSha256(fmt.Sprintf("%v:%v", data, a.Config().EmailSettings.InviteSalt))
-			bodyPage.Props["Link"] = fmt.Sprintf("%s/signup_user_complete/?d=%s&h=%s", siteURL, url.QueryEscape(data), url.QueryEscape(hash))
+
+			if result := <-a.Srv.Store.Token().Save(token); result.Err != nil {
+				l4g.Error(utils.T("api.team.invite_members.send.error"), result.Err)
```
Comment (crspeller) [inline app/email.go]: If we can't save the token, maybe we shouldn't send the email? Since it certainly won't work.

File: app/user.go [hunk 5]
```diff
@@ -43,26 +45,31 @@ func (a *App) CreateUserWithHash(user *model.User, hash string, data string) (*m
 		return nil, err
 	}
 
-	props := model.MapFromJson(strings.NewReader(data))
+	result := <-a.Srv.Store.Token().GetByToken(hash)
```
Comment (crspeller) [inline app/user.go]: Probably should rename the parameter `hash` to `tokenStr` or somthing.

Actually the function needs to be renamed. `CreateUserWithToken`?

File: app/team.go [hunk 6]
```diff
@@ -216,19 +220,25 @@ func (a *App) AddUserToTeamByTeamId(teamId string, user *model.User) *model.AppE
 	}
 }
 
-func (a *App) AddUserToTeamByHash(userId string, hash string, data string) (*model.Team, *model.AppError) {
-	props := model.MapFromJson(strings.NewReader(data))
+func (a *App) AddUserToTeamByToken(userId string, tokenId string, data string) (*model.Team, *model.AppError) {
+	result := <-a.Srv.Store.Token().GetByToken(tokenId)
+	if result.Err != nil {
+		return nil, model.NewAppError("JoinUserToTeamByHash", "api.user.create_user.signup_link_invalid.app_error", nil, result.Err.Error(), http.StatusBadRequest)
```
Comment (saturninoabril) [inline app/team.go]: Should the error be "AddUserToTeamByToken"?

File: app/email.go [hunk 7]
```diff
@@ -270,15 +270,22 @@ func (a *App) SendInviteEmails(team *model.Team, senderName string, invites []st
 			bodyPage.Html["ExtraInfo"] = utils.TranslateAsHtml(utils.T, "api.templates.invite_body.extra_info",
 				map[string]interface{}{"TeamDisplayName": team.DisplayName, "TeamURL": siteURL + "/" + team.Name})
 
+			token := model.NewToken(
+				TOKEN_TYPE_TEAM_INVITATION,
+				model.MapToJson(map[string]string{"team": team.Id, "email": invite}),
```
Comment (saturninoabril) [inline app/email.go]: Maybe use teamId in `"teamId": team.Id`?

File: app/team.go (hunk 6 above)
Comment (jespino) [reply app/team.go]: Yes

File: app/email.go (hunk 7 above)
Comment (jespino) [reply app/email.go]: Yes

File: app/user.go (hunk 5 above)
Comment (jespino) [reply app/user.go]: I create a new CreateUserWithToken but maintain the last one calling the new one for "backward compatibility", I don't know if anybody is using it. If you prefer it, I can remove the old function.

File: model/client4.go [hunk 8]
```diff
@@ -543,6 +543,12 @@ func (c *Client4) CreateUserWithHash(user *User, hash, data string) (*User, *Res
 	}
 }
 
+// CreateUserWithHash creates a user in the system based on the provided user struct and hash created.
```
Comment (crspeller) [inline model/client4.go]: It won't actually work with the old hash. You need to supply a token. So not sure how valuable this backwards compatibility is. Should at least update the comment to warn this.

File: model/client4.go [hunk 9]
```diff
@@ -526,13 +526,13 @@ func (c *Client4) CreateUser(user *User) (*User, *Response) {
...
 
-// CreateUserWithHash creates a user in the system based on the provided user struct and hash created.
-func (c *Client4) CreateUserWithHash(user *User, hash, data string) (*User, *Response) {
+// CreateUserWithToken creates a user in the system based on the provided user struct and tokenId created.
+func (c *Client4) CreateUserWithToken(user *User, tokenId, data string) (*User, *Response) {
 	var query string
-	if hash != "" && data != "" {
-		query = fmt.Sprintf("?d=%v&h=%v", url.QueryEscape(data), hash)
+	if tokenId != "" && data != "" {
+		query = fmt.Sprintf("?d=%v&h=%v", url.QueryEscape(data), tokenId)
```
Comment (crspeller) [inline model/client4.go]: Change `h` to `t`?

File: api/team.go [hunk 10]
```diff
@@ -182,15 +182,15 @@ func removeUserFromTeam(c *Context, w http.ResponseWriter, r *http.Request) {
 
 func addUserToTeamFromInvite(c *Context, w http.ResponseWriter, r *http.Request) {
 	params := model.MapFromJson(r.Body)
-	hash := params["hash"]
+	tokenId := params["hash"]
```
Comment (crspeller) [inline api/team.go]: Can we change "hash" to "token"?

**Comment by jespino** [comment]
Changes made, and the PRs submitted to mattermost/mattermost-redux#463 and mattermost/mattermost-webapp#1109

**Comment by jwilander** [comment]
@jespino will this invalidate all existing invitations on upgrade? E.g. if on 4.9 I send out some invitations then upgrade the next day to 4.10, will those invites break? If so, we should changelog this and let the PMs know

**Comment by jespino** [comment]
@jwilander Yes, it invalidates all existing invitations. We can maintain backward compatibility, but I'm not sure it worth it.

**Comment by jasonblais** [comment]
@jespino Does this **only** invalidate email invite links? But not the team invite link from the "Get Team Invite Link" dialog, nor password reset or email verification links?

**Comment by jespino** [comment]
@jasonblais, exactly, only email invites will be invalidated.

**Comment by jasonblais** [comment]
@jespino Sounds good - backwards compatibility not needed since email invite links expire after 48 hours anyways.

Added as an upgrade notice for 4.10: https://github.com/mattermost/docs/pull/1943

**Pull request merged**